Support localization in static embeds

WiNloSt opened the pull request:
> [!note]
> The BE code is cherry-picked from https://github.com/metabase/metabase/pull/49679, which is already approved, so we don't need extra reviews here.
Please [see this convo for more details.](https://metaboat.slack.com/archives/C064EB1UE5P/p1730810881811109)

[Project doc](https://www.notion.so/metabase/Support-language-localization-in-static-embedding-0307665e3008435d9bb2b3fff8148d7d)

Closes https://github.com/metabase/metabase/issues/8490

### Description

- Set embed locale
- Pass `X-Metabase-Locale` to download APIs. Needs #49531 to work properly.
- Makes the static embeds and public links work with `#locale=xxx`

### How to verify
- Embed various cards with various chart types on a dashboard and test with a static embedding or a public link, I used the dashboard in the example collection.
- pass `#locale=<your_locale>` e.g. `#locale=ko` (Korean) 

### Demo
![Screenshot 2024-11-04 at 10 46 55 PM](https://github.com/user-attachments/assets/f8e5d969-a8ae-4cc5-aa19-f6df63d9c604)

#### Cypress tests
- ![Screenshot 2024-11-05 at 10 57 15 PM](https://github.com/user-attachments/assets/59bb8061-2de8-48e5-b80a-de84a8c6f384)
- ![Screenshot 2024-11-06 at 2 56 00 PM](https://github.com/user-attachments/assets/8dd2278b-099c-4428-a044-7ef47c46e2e7)

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

npretto committed:
wip locale header middleware

WiNloSt committed:
Correct embedding hash options type semantically

WiNloSt committed:
Pass `X-Metabase-Locale` to embed downloads

WiNloSt committed:
Merge branch 'master' into 8490-support-localization-in-static-embeds

WiNloSt committed:
Allow smart scalar text to be translated

WiNloSt commented on frontend/src/metabase/public/LocaleProvider.tsx:
```diff
@@ -15,7 +15,10 @@ export const LocaleProvider = ({
-      loadLocalization(locale).then(() => setIsLoadingLocale(false));
+      loadLocalization(locale).then(translatedObject => {
+        setIsLoadingLocale(false);
+        window.MetabaseUserLocalization = translatedObject;
```
![Screenshot 2024-11-04 at 10 46 55 PM](https://github.com/user-attachments/assets/74b52f2a-8d9b-498c-b920-838bee080ee7)

Given this situation, an instance with locale Chinese, `#locale=ko` (Korean)

There's a logic in `i18n.js` that will change the ttag locale to instance locale then back to users locale. This messes up the locale set by calling `loadLocalization` (Korean) since `window.MetabaseUserLocalization` would still be the same value as the instance locale (Chinese) since there's no user in static embed or public link.

It makes sense to set `window.MetabaseUserLocalization` to the downloaded locale in this case, since this represents the user locale set by the hash option.

WiNloSt commented on frontend/src/metabase/public/LocaleProvider.tsx:
```diff
@@ -15,7 +15,10 @@ export const LocaleProvider = ({
       setIsLoadingLocale(true);
       setLocaleHeader(locale);
-      loadLocalization(locale).then(() => setIsLoadingLocale(false));
+      loadLocalization(locale).then(translatedObject => {
```
This PRs basically does 3 things.
1. Uses `LocaleProvider` on static embeds/public links components.
2. Cherry pick BE change needs to translate some strings
3. Modify this file to make some more strings work.

npretto commented on frontend/src/metabase/public/lib/types.ts:
```diff
@@ -36,6 +36,14 @@ export type EmbeddingDisplayOptions = {
   downloads: boolean | null;
 };
 
+export type AdditionalEmbeddingHashOptions = {
```
Can you help me understand what is "additional" in this context and why these two are not in `EmbeddingDisplayOptions`?

WiNloSt replied:
- embedding display options are the ones we have in static embed modal display options.
    ![image](https://github.com/user-attachments/assets/c08fb43b-5412-4bce-8cdd-c830c767baef)
- dashboards have their own hash options that don't apply to questions `refresh` and `fullscreen` that are used to control the dashboard state.
     https://github.com/metabase/metabase/blob/e840233a4f950d0180333c213fa8b6478b391852/frontend/src/metabase/dashboard/types/hash-options.ts#L3-L11
- Both questions and dashboards have these common hash options that's why I called them `AdditionalEmbeddingHashOptions`

npretto replied:
Ah thanks, can we add some small comments about this in the code?

WiNloSt replied:
Sure thing.

WiNloSt replied:
Added in https://github.com/metabase/metabase/pull/49441/commits/5dd8fd3067d53431fd44b1050872226ba6ba5b4a

WiNloSt committed:
Merge branch 'master' into 8490-support-localization-in-static-embeds

WiNloSt committed:
Stop using global variable

WiNloSt committed:
Add more details to embedding types

WiNloSt commented on frontend/src/metabase/public/containers/PublicOrEmbeddedDashboard/PublicOrEmbeddedDashboard.tsx:
```diff
@@ -208,33 +212,35 @@ const PublicOrEmbeddedDashboardInner = ({
-      cardTitled={cardTitled}
-      downloadsEnabled={downloadsEnabled}
-    />
+    <LocaleProvider locale={locale}>
```
I deliberately pass `locale` from the parent component so I could have `LocaleProvider` wrapped `PublicOrEmbeddeDashboardView` similar to how I also have it wrapped `PublicOrEmbeddedQuestionView`.

I could use `LocaleProvider` in the parent component, but I think putting it here would make it easier to compare between the 2 similar view components.

npretto replied:
The important thing is that it's rendered right away (not after a loading) so that it fires the requests asap. It looks that where it is now it works like that so it's fine

WiNloSt commented on frontend/src/metabase/lib/i18n.js:
```diff
@@ -32,6 +32,8 @@ export async function loadLocalization(locale) {
         };
   setLocalization(translationsObject);
+
+  return translationsObject;
```
This is used in `frontend/src/metabase/public/LocaleProvider.tsx`. Since `withInstanceLanguage` function resets the locale, but in static embeddings/public links, there is no user locale so `MetabaseUserLocalization` is set to the instance locale, so after the function has called, the locale that has been loaded by `loadLocalization` will be overridden by the instance locale.

This return value is used to set the `MetabaseUserLocalization` when necessary. For now it's only used in `LocaleProvider` like mentioned previously.

dpsutton commented:
I don't think this BE change is necessary. I'm using `master` and changing the dashboard embed endpoint to just return a string that is translateable and the language:


```clojure
(api/defendpoint GET "/dashboard/:token"
  "Fetch a Dashboard via a JSON Web Token signed with the `embedding-secret-key`.

   Token should have the following format:

     {:resource {:dashboard <dashboard-id>}}"
  [token]
  (let [unsigned (unsign-and-translate-ids token)]
    {:hello (metabase.util.i18n/tru "Database syncing")
     :locale (.getDisplayLanguage (metabase.util.i18n/user-locale))}))
```

```shell
❯ http get 'http://localhost:3000/api/embed/dashboard/eyJhbGciOiJIUzI1NiJ9.eyJyZXNvdXJjZSI6eyJkYXNoYm9hcmQiOjEyfSwicGFyYW1zIjp7fX0.QZ-tEWpgJeSMUuaPfuH6_VQwM2yXNMcuureWCtHsszU' x-metabase-locale:fr -pb
{
    "hello": "Synchronisation de la base de données",
    "locale": "French"
}

❯ http get 'http://localhost:3000/api/embed/dashboard/eyJhbGciOiJIUzI1NiJ9.eyJyZXNvdXJjZSI6eyJkYXNoYm9hcmQiOjEyfSwicGFyYW1zIjp7fX0.QZ-tEWpgJeSMUuaPfuH6_VQwM2yXNMcuureWCtHsszU' x-metabase-locale:zh -pb
{
    "hello": "数据库正在同步",
    "locale": "Chinese"
}
```

WiNloSt commented:
So almost everything already works except chart labels.
#### Without a patch
![Screenshot 2024-11-05 at 9 43 58 PM](https://github.com/user-attachments/assets/588ee4cc-e4bb-4137-ba3e-22b6e90b80b3)

#### With a patch
![Screenshot 2024-11-05 at 9 45 32 PM](https://github.com/user-attachments/assets/8f300459-9080-4ee8-8e49-0c132a74ff50)

The endpoint that's called
`http://localhost:3000/api/public/dashboard/8243122d-5d6b-4fb4-b154-09b929ef54fe/dashcard/21/card/12?parameters=%5B%7B%22type%22%3A%22string%2F%3D%22%2C%22value%22%3Anull%2C%22id%22%3A%22fc2cd1be%22%2C%22target%22%3A%5B%22dimension%22%2C%5B%22field%22%2C60%2C%7B%22base-type%22%3A%22type%2FText%22%2C%22source-field%22%3A40%7D%5D%5D%7D%2C%7B%22type%22%3A%22date%2Frange%22%2C%22value%22%3Anull%2C%22id%22%3A%22afa56954%22%2C%22target%22%3A%5B%22dimension%22%2C%5B%22field%22%2C41%2C%7B%22base-type%22%3A%22type%2FDateTime%22%7D%5D%5D%7D%2C%7B%22type%22%3A%22string%2F%3D%22%2C%22value%22%3Anull%2C%22id%22%3A%225eeec658%22%2C%22target%22%3A%5B%22dimension%22%2C%5B%22field%22%2C58%2C%7B%22base-type%22%3A%22type%2FText%22%2C%22source-field%22%3A40%7D%5D%5D%7D%2C%7B%22type%22%3A%22string%2F%3D%22%2C%22value%22%3Anull%2C%22id%22%3A%22512c560a%22%2C%22target%22%3A%5B%22dimension%22%2C%5B%22field%22%2C48%2C%7B%22base-type%22%3A%22type%2FText%22%2C%22source-field%22%3A43%7D%5D%5D%7D%5D`

WiNloSt committed:
Extend the E2E util type and correct some properties types

WiNloSt committed:
Add e2e test

WiNloSt committed:
Merge branch 'master' into 8490-support-localization-in-static-embeds

WiNloSt commented on e2e/test/scenarios/embedding/embedding-reproductions.cy.spec.js:
```diff
@@ -1024,3 +1025,71 @@ describe.skip("issue 49142", () => {
+      );
+    });
+
+    cy.findByTestId("embed-frame").within(() => {
```
![Screenshot 2024-11-05 at 10 57 15 PM](https://github.com/user-attachments/assets/f284bea1-9a27-4a09-8109-4dd7e18a2742)

WiNloSt commented on e2e/test/scenarios/embedding/embedding-reproductions.cy.spec.js:
```diff
@@ -1024,3 +1025,71 @@ describe.skip("issue 49142", () => {
+      cy.findByText("제공:").should("be.visible");
+
+      // Aggregation "count"
+      cy.findByText("카운트").should("be.visible");
```
This assertion on the chart label (aggregation) will fail without [the BE patch](https://github.com/metabase/metabase/pull/49441/commits/83d8d1358ce30019291b5a0b17d69528f196227a).

You can try to revert that commit and run this test again, and the test should fail (I've tested that already)

npretto replied:
I was going to suggest if we can use a question/dashboard already present in the test db, but I guess we need the aggregation specifically for this assertion

WiNloSt replied:
I think the difference would be minimal, wouldn't it? With a few less requests for creating a question and a dashboard. That might save like 150ms on Cypress on my Machine (could be longer on CI I guess), but I couldn't find a compatible dashboard that alreay exist, so I guess it's fine to do so.

And we've been creating new resources like this in a lot of tests anyway.

Why do you think we should use existing data? Was this concern raised before?

npretto replied:
I just find the code to create the questions un-necessary if we already have the questions pre-made, but it's probably personal preference.
Creating them ad-hoc probably makes it clearer what we're opening

WiNloSt replied:
there are pros and cons to using premade questions.
**Pros**
- Faster. Maybe this could be a huge boost imagine that we have a pool of questions and dashboards that maybe work for like 80% of the use cases. Then we could potentially reduce the time to run tests, hopefully, by a lot.

**Cons**
- Dependent on the DB snapshot
- Less clear what question we're referring to. Compare to the immediate `questionDetails` that you could read in each test.

Again unless we know exactly what we're trying to achieve here. I'd still prefer creating new resources for redability.

WiNloSt commented on e2e/test/scenarios/embedding/embedding-reproductions.cy.spec.js:
```diff
@@ -1024,3 +1025,71 @@ describe.skip("issue 49142", () => {
+        },
+        {
+          additionalHashOptions: {
+            locale: "ko",
```
Korean

WiNloSt commented on frontend/src/metabase/public/LocaleProvider.tsx:
```diff
@@ -15,7 +15,10 @@ export const LocaleProvider = ({
-      loadLocalization(locale).then(() => setIsLoadingLocale(false));
+      loadLocalization(locale).then(translatedObject => {
+        setIsLoadingLocale(false);
+        setUserLocale(translatedObject);
```
See this [original comment](https://github.com/metabase/metabase/pull/49441#discussion_r1827967395) on why I did this.

npretto reviewed (approved):
Code looks good to me 🙏

WiNloSt committed:
Fix dashboard name

WiNloSt committed:
Merge branch 'master' into 8490-support-localization-in-static-embeds

WiNloSt commented on frontend/src/metabase/public/lib/types.ts:
```diff
@@ -36,6 +39,19 @@ export type EmbeddingDisplayOptions = {
+ * Properties here exists only in the document (just `hide_parameters` since `locale` is a new one),
+ * but not in the UI.
+ */
+export type AdditionalEmbeddingHashOptions = {
```
Yeah. I actually like having the namespace first since it makes searching easier.

WiNloSt replied:
Fixed in https://github.com/metabase/metabase/pull/49441/commits/b15a382e68ee6ae8de29f1b33086b4324a5a7f5e

WiNloSt committed:
Add a test for static embed question

WiNloSt committed:
Make a type name follows a convention

Co-authored-by: Denis Berezin <denis.berezin@metabase.com>

WiNloSt committed:
Fix failed E2E tests due to missing default values

WiNloSt commented on e2e/test/scenarios/embedding/embedding-reproductions.cy.spec.js:
```diff
@@ -1024,3 +1025,100 @@ describe.skip("issue 49142", () => {
+      );
+    });
+
+    cy.findByTestId("embed-frame").within(() => {
```
Question test
![Screenshot 2024-11-06 at 2 56 00 PM](https://github.com/user-attachments/assets/8dd2278b-099c-4428-a044-7ef47c46e2e7)

WiNloSt commented on e2e/test/scenarios/embedding/embedding-reproductions.cy.spec.js:
```diff
@@ -1024,3 +1025,100 @@ describe.skip("issue 49142", () => {
+
+    cy.findByTestId("embed-frame").within(() => {
+      // X-axis labels: Jan 2023
+      cy.findByText("4월 2022").should("be.visible");
```
Change the assertion to the first visible x-axis label.

WiNloSt committed:
Fix public embed not working with `?locale` anymore

WiNloSt committed:
Merge branch 'master' into 8490-support-localization-in-static-embeds

dpsutton committed:
Preserve locale when in the `as-admin` macro is used

Motivating issue:

```shell
❯ http get 'http://localhost:3000/api/embed/dashboard/eyJhbGciOiJIUzI1NiJ9.eyJyZXNvdXJjZSI6eyJkYXNoYm9hcmQiOjEyfSwicGFyYW1zIjp7fX0.QZ-tEWpgJeSMUuaPfuH6_VQwM2yXNMcuureWCtHsszU/dashcard/162/card/124' x-metabase-locale:fr -pb | jq '.data|.cols|map(.display_name)'
[
  "State",
  "Count"
]
```

I'm running a query that has aggregated columns and it is returning the
untranslated. Note that these are computed and dynamically given a
display name.

```clojure
(case tag
  :count     (i18n/tru "Count")
  :cum-count (i18n/tru "Cumulative count"))
```

The problem is that we run embedding things as admin. And that gets the
current user, which in embed and public spaces is nil, and so binds a
nil to user-locale, which defaults to english when translating.

After:

```
❯ http get 'http://localhost:3000/api/embed/dashboard/eyJhbGciOiJIUzI1NiJ9.eyJyZXNvdXJjZSI6eyJkYXNoYm9hcmQiOjEyfSwicGFyYW1zIjp7fX0.QZ-tEWpgJeSMUuaPfuH6_VQwM2yXNMcuureWCtHsszU/dashcard/162/card/124' x-metabase-locale:fr -pb | jq '.data|.cols|map(.display_name)'
[
  "State",
  "Nombre de lignes"
]

❯ http get 'http://localhost:3000/api/embed/dashboard/eyJhbGciOiJIUzI1NiJ9.eyJyZXNvdXJjZSI6eyJkYXNoYm9hcmQiOjEyfSwicGFyYW1zIjp7fX0.QZ-tEWpgJeSMUuaPfuH6_VQwM2yXNMcuureWCtHsszU/dashcard/162/card/124' x-metabase-locale:zh -pb | jq '.data|.cols|map(.display_name)'
[
  "State",
  "行数"
]

❯ http get 'http://localhost:3000/api/embed/dashboard/eyJhbGciOiJIUzI1NiJ9.eyJyZXNvdXJjZSI6eyJkYXNoYm9hcmQiOjEyfSwicGFyYW1zIjp7fX0.QZ-tEWpgJeSMUuaPfuH6_VQwM2yXNMcuureWCtHsszU/dashcard/162/card/124' x-metabase-locale:es -pb | jq '.data|.cols|map(.display_name)'
[
  "State",
  "Contar"
]
```

This is fetching a dashboard's card which is a query of orders grouped
by created at month, sum of total.

WiNloSt committed:
Revert "wip locale header middleware"

This reverts commit 83d8d1358ce30019291b5a0b17d69528f196227a.

WiNloSt committed:
Merge branch 'master' into 8490-support-localization-in-static-embeds
